src/sentry/debug_files/artifact_bundle_indexing.py

```diff
@@ -227,70 +171,146 @@ def get_flat_file_meta(self) -> Optional[FlatFileMeta]:
+        # and throw an error otherwise which we will pass down and handle in the caller.
+        try:
+            flat_file_index = (
+                ArtifactBundleFlatFileIndex.objects.filter(
+                    project_id=identifier.project_id,
+                    release_name=identifier.release,
+                    dist_name=identifier.dist,
+                )
+                .select_for_update(nowait=not blocking)
+                .first()
+            )
+        except DatabaseError:
+            return False
 
         index = FlatFileIndex()
         # Load the index from the file if it exists
         if existing_index := flat_file_index.load_flat_file_index():
             index.from_json(existing_index)
 
-        # Before merging new data into the index, we will clear any existing
-        # data from the index related to this bundle.
-        # This is related to an edge-case in which the same `bundle_id` could be
-        # re-used but with different file contents.
-        index.remove(bundle_meta.id)
+        for bundle in bundles_to_add or []:
+            # Before merging new data into the index, we will clear any existing
+            # data from the index related to this bundle.
+            # This is related to an edge-case in which the same `bundle_id` could be
+            # re-used but with different file contents.
+            index.remove(bundle.meta.id)
 
-        # We merge the index based on the identifier type.
-        if identifier.is_indexing_by_release():
-            index.merge_urls(bundle_meta, bundle_archive)
-        else:
-            index.merge_debug_ids(bundle_meta, bundle_archive)
+            # We merge the index based on the identifier type.
+            if identifier.is_indexing_by_release():
+                index.merge_urls(bundle.meta, bundle.urls)
+            else:
+                index.merge_debug_ids(bundle.meta, bundle.debug_ids)
+
+        for bundle_id in bundles_to_remove or []:
+            index.remove(bundle_id)
 
         # Store the updated index file
         new_json_index = index.to_json()
         flat_file_index.update_flat_file_index(new_json_index)
 
-        # And then mark the bundle as indexed
-        was_updated = FlatFileIndexState.compare_state_and_set(
-            flat_file_index.id,
-            bundle_meta.id,
-            ArtifactBundleIndexingState.NOT_INDEXED,
-            ArtifactBundleIndexingState.WAS_INDEXED,
-        )
-        if not was_updated:
-            metrics.incr("artifact_bundle_flat_file_indexing.duplicated_indexing")
-            logger.error("`ArtifactBundle` %r was already indexed into %r", bundle_meta, identifier)
+        # And then mark the bundles as indexed
+        for bundle in bundles_to_add or []:
+            was_updated = FlatFileIndexState.mark_as_indexed(
+                flat_file_index_id=flat_file_index.id, artifact_bundle_id=bundle.meta.id
+            )
+            if not was_updated:
```

Comment:
I am curious to know how this can happen, given the `SELECT FOR UPDATE`, will be curious to check the metrics in production.

Reply:
In the followup PR, the backfill task is split into two phases to keep the lock as short as possible:
https://github.com/getsentry/sentry/pull/54541/files#diff-496e4699a249aca7a8a1027116f613f9260e650007c69549a80a99a3ee7ef97dR276-R279

1. prepare everything we need to backfill into the index.
2. actually perform the read-modify-write behind the lock.

In theory there can be races here and we are double-indexing. Although the index update itself is idempotent so this does not cause any problems except maybe a bit of duplicated work. I’m also curious how often we will be hitting this in practice.